src/openms/include/OpenMS/CHEMISTRY/DigestionEnzymeDB.h

```diff
@@ -0,0 +1,288 @@
+// $Maintainer: Xiao Liang $
+// $Authors: Xiao Liang, Chris Bielow $
+// --------------------------------------------------------------------------
+
+#ifndef OPENMS_CHEMISTRY_DIGESTIONENZYMEDB_H
+#define OPENMS_CHEMISTRY_DIGESTIONENZYMEDB_H
+
+#include <OpenMS/CHEMISTRY/DigestionEnzyme.h>
+#include <OpenMS/CONCEPT/LogStream.h>
+#include <OpenMS/DATASTRUCTURES/Map.h>
+#include <OpenMS/DATASTRUCTURES/String.h>
+#include <OpenMS/FORMAT/ParamXMLFile.h>
+#include <OpenMS/SYSTEM/File.h>
+
+#include <boost/unordered_map.hpp>
+#include <set>
+
+namespace OpenMS
+{
+  /**
+    @ingroup Chemistry
+
+    @brief Digestion enzyme database (base class)
+
+    Template parameters:
+    @p DigestionEnzymeType should be a subclass of DigestionEnzyme.
+    @p InstanceType should be a subclass of DigestionEnzymeDB ("Curiously Recurring Template Pattern", see https://stackoverflow.com/a/34519373).
+  */
+  template<typename DigestionEnzymeType, typename InstanceType> class OPENMS_DLLAPI DigestionEnzymeDB
+  {
+  public:
+
+    /** @name Typedefs
+    */
+    //@{
+    typedef typename std::set<const DigestionEnzymeType*>::const_iterator ConstEnzymeIterator;
+    typedef typename std::set<const DigestionEnzymeType*>::iterator EnzymeIterator;
+    //@}
+
+    /// this member function serves as a replacement of the constructor
+    static InstanceType* getInstance()
+    {
+      static InstanceType* db_ = 0;
+      if (db_ == 0)
+      {
+        db_ = new InstanceType;
+      }
+      return db_;
+    }
+
+    /** @name Constructors and Destructors
+    */
+    //@{
+    /// destructor
+    virtual ~DigestionEnzymeDB()
+    {
+      clear();
+    }
+    //@}
+
+    /** @name Accessors
+    */
+    //@{
+    /// returns a pointer to the enzyme with name (supports synonym names)
+    /// @throw Exception::ElementNotFound if enzyme is unknown
+    /// @note enzymes are registered in regular and in toLowercase() style, if unsure use toLowercase
+    const DigestionEnzymeType* getEnzyme(const String& name) const
+    {
+      if (enzyme_names_.find(name) == enzyme_names_.end())
+      {
+        throw Exception::ElementNotFound(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION, name);
+      }
+      return enzyme_names_.at(name);
+    }
+
+    /// returns a pointer to the enzyme with cleavage regex
+    /// @throw Exception::IllegalArgument if enzyme regex  is unregistered.
+    const DigestionEnzymeType* getEnzymeByRegEx(const String& cleavage_regex) const
+    {
+      if (!enzyme_regex_.has(cleavage_regex))
+      {
+        // @TODO: why does this use a different exception than "getEnzyme"?
+        throw Exception::IllegalArgument(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION,
+                                         String("Enzyme with regex " + cleavage_regex + " was not registered in Enzyme DB, register first!").c_str());
+      }
+      return enzyme_regex_[cleavage_regex];
+    }
+
+    /// load enzymes from given file
+    void setEnzymes(const String& filename)
```

Comment:
shouldn't we make the DB read only? Otherwise we will need a lot of looking

Reply:
That function was already there, but you make a valid point. It's not actually used anywhere in OpenMS and I'm happy to remove it.